### DISCUSSION TITLE
explicit return values, why and how does this work

### DISCUSSION THREAD

**Discussion by koin612** [YuriGor/deepdash#92] [Q&A] [answered]
hey, 

I used deepdash's filterdeep method and wrote a predicate which evaluates a boolean expression like

`(val, key, parent) => key === 'key' && (val === 'val1' || val === 'val2')`

this doesnt work the result would be null, but if I explicit return true everything is fine (like in the official examples).

`(val, key, parent) => { if (...) return true }`

So, from the docs (https://deepdash.io/#iteratee) I figured out that this is because of "returns - return false explicitly to prevent iteration over current value's children"?

Is my assumption correct and what is this called in JS? I'd like to know more about this but I fail to find anything related to this. Can someone explain or point out some words so I can check?

Edit: I now think it's because it works with undefined for seen nodes so just returning true/false would not work?

**Comment by YuriGor** [comment, marked as answer]
Hi, looks like you mixing docs for eachDeep and filterDeep.

[eachDeep](https://deepdash.io/#eachdeep-foreachdeep) will not care about returned value unless it's exactly boolean `false`. 
In this case eachDeep will skip current children if any, so it's kind of 'break' in for loop, but local for current depth level.
This is similar to convention of Lodash [each](https://lodash.com/docs/4.17.15#forEach) method.

[filterDeep](https://deepdash.io/#filterdeep) similar to [Lodash](https://lodash.com/docs/4.17.15#filter) or native [array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) filter method expects true or false from predicate to decide keep current value or not. 
Because filterDeep case is much more complex, I extended this convention and also accept `undefined` as possible return result. If "truthy" reply will be "yes" and "falsey" response will be "no", `undefined` means "Don't know yet"

In case of `undefined` current value will be preserved in result only in case some children values will also be preserved. If finally in result such value will be childless - it will be cleared out.
`filterDeep` deep doesn't care too much about true or false result data type, 0 or null will also work as false. It only cares about `undefined` as separate from `false` reply, so `undefined` is not falsey from filterDeep point of view.

If you are not sure yet you can read about [Truthy](https://developer.mozilla.org/en-US/docs/Glossary/Truthy) and [Falsy](https://developer.mozilla.org/en-US/docs/Glossary/Falsy) values in JS.

Note: if in some case you function will not return something - it will mean function will still return `undefined` implicitly.
I guess this is your point of confusion. So you can of coarse return just result of evaluation of some boolean value, you don't need to return exactly `true` or `false`, in case of eachDeep just make sure your false result is really boolen, not `undefined` or `0` wich in boolean expressions are considered as false, but if you check it with `===` operator - it will be different.
And in case of filterDeep you can return any truthy/falsy values, but be carefull with `undefined` because it's a separate reply.

To have better idea how deepdash works with complex nested data - I recomment first to learn how Lodash / JS array `forEach` and `filter` methods work. They work similar, but Lodash also supports objects, not only arrays, and in Lodash there is return false ==> break convention exists, while native JS forEach doesn't support it.

**Reply by koin612** [reply]
Hey YuriGor,

thank for your detailed explanation, really appreciate it! 
I think I got messed up by the definition of a predicate. For me it's like `P: X -> {true,false}` and so I got stuck with that thinking and messed up also understanding the documentation. 

But I think I got it now and it makes perfectly sense :)

**Reply by YuriGor** [reply]
Sure, documentation is not strongest side of this lib, so if you have some motivation / resources - feel free to improve docs, I am always open for PR.
docs sources live here: https://github.com/YuriGor/deepdash/tree/master/docs
